rosbag2/include/rosbag2/reader.hpp

```diff
@@ -0,0 +1,133 @@
+#ifdef _WIN32
+# pragma warning(push)
+# pragma warning(disable:4251)
+#endif
+
+namespace rosbag2
+{
+/**
+ * The Reader allows opening and reading messages of a bag. Messages will be read
+ * sequentially according to timestamp.
+ */
+class ROSBAG2_PUBLIC Reader
+{
+public:
+  explicit Reader(
+    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory =
+    std::make_unique<rosbag2_storage::StorageFactory>(),
+    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory =
+    std::make_shared<SerializationFormatConverterFactory>());
+
+  virtual ~Reader();
+
+  void reset();
+
+  /**
+   * Open a rosbag for reading messages sequentially (time-ordered). Throws if file could not be
+   * opened. This must be called before any other function is used. The rosbag is
+   * automatically closed on destruction.
+   *
+   * If the `output_serialization_format` within the `converter_options` is not the same as the
+   * format of the underlying stored data, a converter will be used to automatically convert the
+   * data to the specified output format.
+   * Throws if the converter plugin does not exist.
+   *
+   * \param storage_options Options to configure the storage
+   * \param converter_options Options for specifying the output data format
+   */
+  virtual void open(
+    const StorageOptions & storage_options, const ConverterOptions & converter_options);
+
+  /**
+   * Ask whether the underlying bagfile contains at least one more message.
+   *
+   * \return true if storage contains at least one more message
+   * \throws runtime_error if the Reader is not open.
+   */
+  virtual bool has_next();
+
+  /**
+   * Read next message from storage. Will throw if no more messages are available.
+   * The message will be serialized in the format given to `open`.
+   *
+   * Expected usage:
+   * if (reader.has_next()) message = reader.read_next();
+   *
+   * \return next message in serialized form
+   * \throws runtime_error if the Reader is not open.
+   */
+  virtual std::shared_ptr<SerializedBagMessage> read_next();
+
+  /**
+   * Ask bagfile for all topics (including their type identifier) that were recorded.
+   *
+   * \return vector of topics with topic name and type as std::string
+   * \throws runtime_error if the Reader is not open.
+   */
+  virtual std::vector<TopicMetadata> get_all_topics_and_types();
+
+  /**
+   * Ask if we have a storage object to read from.
+   *
+   * \return true if storage exists. false if it is null.
+   */
+  bool has_storage();
+
+  /**
+   * Open a new database file to read from.
+   *
+   * \return true if there is a file to read from. false otherwise.
+   */
+  bool open_read_only(const std::string & file, const std::string & storage_id);
+
+  rosbag2_storage::BagMetadata get_metadata();
+
+private:
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory_;
+  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_;
```

Comment:
@Karsten1987 might have more insight into this. 
If I understood this well, there can only be one storage object open at a time to avoid multiple objects reading from a database.

Reply:
I had more time to read the code tonight and it seems that we should change the factories to return `std::unique_ptr`, but this is for another PR...

Rationale: https://stackoverflow.com/questions/37884728/does-c11-unique-ptr-and-shared-ptr-able-to-convert-to-each-others-type

> std::unique_ptr is the C++11 way to express exclusive ownership, but one of its most attractive features is that it easily and efficiently converts to a std::shared_ptr.
> 
> This is a key part of why std::unique_ptr is so well suited as a factory function return type. Factory functions can’t know whether callers will want to use exclusive ownership semantics for the object they return or whether shared ownership (i.e., std::shared_ptr) would be more appropriate. By returning a std::unique_ptr, factories provide callers with the most efficient smart pointer, but they don’t hinder callers from replacing it with its more flexible sibling.
> 
> std::shared_ptr to std::unique_ptr is not allowed. Once you’ve turned lifetime management of a resource over to a std::shared_ptr, there’s no changing your mind. Even if the reference count is one, you can’t reclaim ownership of the resource in order to, say, have a std::unique_ptr manage it.
> 
> Reference: Effective Modern C++. 42 SPECIFIC WAYS TO IMPROVE YOUR USE OF C++11 AND C++14. Scott Meyers.

tl;dr it is usually better to have factories return unique_ptr, if your user wants to convert to a shared_ptr, it can always happen transparently.